### PR TITLE
Renewal payments are now included in reports (the subsequent pull request)

### LINKED ISSUE #5398: Renewals should appear in the donation report
## User Story

As a [type of user], I want [some goal] so that [some reason].

## Steps to Reproduce

1. Go to...
2. Select...
3. Scroll down to...
4. View error

## System Information

<details><summary>Details</summary>

</details>

## Acceptance Criteria

-   [ ] Renewal donations are included in reports data.
-   [ ] Renewal donations are filtered by `mode` (live or test).

### PR DESCRIPTION
**Pull request by kjohnson**
Resolves #5398

## Description

Renewal payments weren't being included in the results of the `Give_Payments_Query` because they do not include a value for the `_give_payment_mode` meta.

Ideally, the renewal system would be updated to include this meta key/value, but that would require a database update to correct historical data.

Instead, I opted to add an additional query for renewal payments and combine the results.

## Affects

Updates the shared `Endpoint::getPayments()` to combine one-time and renewal payments.

## Visuals

image

## Pre-review Checklist

-   [x] Acceptance criteria satisfied and marked in related issue
-   [x] Relevant `@since` tags included in DocBlocks
-   [x] Changelog updated
-   [x] [Self Review](https://give.gitbook.io/development-manual/devops/github/code-reviews#self-review) of code and UX completed

## Testing Instructions

1. Create a new donation form
1. Create a donation
1. Manually create a renewal
1. See Donations > Reports

Note: Due to #5399, you may need to manually adjust the start date to include first day's payments (for a fresh install).

The renewal payment is now listed in the reports.

### CODE REVIEW

File: src/API/Endpoints/Reports/Endpoint.php [hunk 1]
```diff
@@ -410,13 +410,70 @@ public function getPayments( $startStr, $endStr, $orderBy = 'date', $number = -
...
+				( $this->testMode ) ? 'test' : 'live',
+				$this->currency
+			)
+		);
+		return array_map(
+			function( $renewalPaymentID ) {
+				return new Give_Payment( $renewalPaymentID );
+			},
+			$renewalPaymentIDs
+		);
```
Comment (JasonTheAdams) [inline src/API/Endpoints/Reports/Endpoint.php:474]: I believe this is going to cause an N + 1 problem wherein you are querying all of the payment IDs together, but then querying each payment object individually, therefore for every ID you have 1 additional query. Couldn't you instead query `renewals.*` and then generate the payment (or at least the `WP_Post`) from the data? That way it's all done in a single query.

File: src/API/Endpoints/Reports/Endpoint.php [hunk 2]
```diff
@@ -410,13 +410,70 @@ public function getPayments( $startStr, $endStr, $orderBy = 'date', $number = -
...
+		/**
+		 * Query renewal payments based on parent payments.
+		 * Because the renewal does not have a payment mode stored
+		 * the results are not included in the payment query,
+		 * so we have to run a second query and combine results.
+		*/
+		$payments = array_merge(
+			$payments,
+			$this->getRenewalPayments( $startStr, $endStr, $gateway )
+		);
```
Comment (JasonTheAdams) [inline src/API/Endpoints/Reports/Endpoint.php:422]: I would imagine both queries should be sorted by `DATE DESC`, which case after merging these we would need to do a `usort()` to put them in order.

**Review by JasonTheAdams — CHANGES_REQUESTED** [review]
This approach seems solid. I have a couple notes, however, on what I think could be improved. Open to any feedback, however, if I'm misunderstanding anything.

File: src/API/Endpoints/Reports/Endpoint.php (hunk 1 above)
Comment (kjohnson) [reply src/API/Endpoints/Reports/Endpoint.php:474]: Ah, yup. I see the nested `get_post()` inside the `Give_Payment` definition.

Comment (kjohnson) [reply src/API/Endpoints/Reports/Endpoint.php:474]: @JasonTheAdams the query is quickly getting out of hand, given all the granular data points that the reporting page uses for each payment. I may need to change course and query all donationmeta and then sort in PHP, as opposed to trying to join each of these different meta items in a single query.

**Comment by kevinwhoffman** [comment]
@JasonTheAdams @kjohnson We're close to the 2.9.0 release, and it sounds like we're still figuring out the best approach here, so I'm going to suggest we target the `develop` branch and address this fix in a patch release.

**Comment by kevinwhoffman** [comment]
Per Slack discussion, I changed the base to `develop` and we will target this fix in a patch release following GiveWP 2.9.0.

**Comment by kjohnson** [comment]
I'm going to open a new PR off of `develop` since this branch has `release/2.9.0` changes.

**Pull request closed without merging**